packages/agent/src/agent/http/index.ts

```diff
@@ -407,63 +417,167 @@ export class HttpAgent implements Agent {
+      if (subnetStatus) {
+        resolve(subnetStatus);
+      } else {
+        this.fetchSubnetKeys(canisterId)
+          .then(response => {
+            resolve(response);
+          })
+          .catch(error => {
+            reject(error);
+          });
+      }
     });
+    const [query, subnetStatus] = await Promise.all([queryPromise, subnetStatusPromise]);
+    // Skip verification if the user has disabled it
+    if (!this.#verifyQuerySignatures) {
+      return query;
+    }
+    return this.#verifyQueryResponse(query, subnetStatus);
+  }
 
-    // Apply transform for identity.
-    transformedRequest = await id?.transformRequest(transformedRequest);
+  /**
+   * See https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query for details on validation
+   * @param queryResponse - The response from the query
+   * @param subnetStatus - The subnet status, including all node keys
+   * @returns ApiQueryResponse
+   */
+  #verifyQueryResponse = (
+    queryResponse: ApiQueryResponse,
+    subnetStatus: SubnetStatus | void,
+  ): ApiQueryResponse => {
+    if (!subnetStatus || this.#verifyQuerySignatures === false) {
+      // This should not be called if the user has disabled verification
+      return queryResponse;
+    }
+    const { status, signatures, requestId } = queryResponse;
+
+    const domainSeparator = new TextEncoder().encode('\x0Bic-response');
+    signatures?.forEach(sig => {
+      const { timestamp } = sig;
+      let hash: ArrayBuffer;
+
+      // Hash is constructed differently depending on the status
+      if (status === 'replied') {
+        const { reply } = queryResponse;
+        hash = hashOfMap({
+          status: status,
+          reply: hashOfMap(reply),
+          timestamp: BigInt(timestamp),
+          request_id: requestId,
+        });
+      } else if (status === 'rejected') {
+        const { reject_code, reject_message, error_code } = queryResponse;
+        hash = hashOfMap({
+          status: status,
+          reject_code: reject_code,
+          reject_message: reject_message,
+          error_code: error_code,
+          timestamp: BigInt(timestamp),
+          request_id: requestId,
+        });
+      } else {
+        throw new Error(`Unknown status: ${status}`);
+      }
 
-    const body = cbor.encode(transformedRequest.body);
-    const response = await this._requestAndRetry(() =>
-      this._fetch('' + new URL(`/api/v2/canister/${canister.toText()}/query`, this._host), {
-        ...this._fetchOptions,
-        ...transformedRequest.request,
-        body,
-      }),
-    );
+      const separatorWithHash = concat(domainSeparator, new Uint8Array(hash));
 
-    const queryResponse: QueryResponse = cbor.decode(await response.arrayBuffer());
+      // FIX: check for match without verifying N times
+      const matchingKey = subnetStatus?.nodeKeys.find(key => {
+        const pubKey = new Uint8Array(fromHex(key).slice(12, 44));
```

Comment:
What about defining a method `from_der` somewhere?

Reply:
I got some feedback from Franz-Stefan of the Crypto team. Maybe it is worth adding some more checks on the Der-encoding here.

> most likely, the basis of this decoding is a DER-encoded ASN.1 structure called SubjectPublicKeyInfo defined in [RFC 5280](https://datatracker.ietf.org/doc/html/rfc5280). This structure contains an algorithm identifier, and the public key itself in the subjectPublicKey field. If the public key is an Ed25519 public key, then according to [RFC 8410](https://datatracker.ietf.org/doc/html/rfc8410) the algorithm identifier must be 1.3.101.112.

> While the approach you are using of just extracting the key from the expected location generally works, it would be better if you also ensure that the algorithm ID is really correct. The DER-encoding also carries information of how long the contained subjectPublicKey is, which you could also check, which would give additional assurance that you are dealing with the correct/expected data.

> I don’t know the Javascript world so well, but ideally you would use some ASN.1 DER decoder and make sure all the parts of the DER-encoding have the expected format.

Reply:
I've accomplished this by bringing in our DER-encoding logic for ED25519 public keys from the @dfinity/identity package